### PR TITLE
TIP-480: Define the format standard

### PR DESCRIPTION
**Pull request by skeleton**

### CODE REVIEW

File: STANDARD_FORMAT.md [hunk 1]
```diff
@@ -0,0 +1,946 @@
...
+    
+        array:2 [
+          "a_decimal" => "46546.65987313"
+          "a_negative_deciaml" => "-45.8981226"
+        ]
+
+Linked entities are represented only by their identifier as strings. For instance, the standard format of a *foo* object that has a link to an external *bar* object would be:
+    
+        array:1 [
+          "bar" => "here is the identifier of the bar object"
```
Comment (SamirBoulil) [inline STANDARD_FORMAT.md]: maybe we should say "here_is_the_identifier_of_the_bar_object"

File: STANDARD_FORMAT.md [hunk 2]
```diff
@@ -0,0 +1,946 @@
...
+              "products" => array:1 [
+                0 => "bar"
+              ]
+            ]
+          ]
+        ]
+
+The product values are provided via the key *values*.
+
+Product values can be localisable and/or scopable:
```
Comment (SamirBoulil) [inline STANDARD_FORMAT.md]: localizable

File: STANDARD_FORMAT.md [hunk 3]
```diff
@@ -0,0 +1,946 @@
...
+
+* *localisable* means its value depends of the locale
```
Comment (SamirBoulil) [inline STANDARD_FORMAT.md]: same

Comment (SamirBoulil) [inline STANDARD_FORMAT.md]: depends on

File: STANDARD_FORMAT.md [hunk 4]
```diff
@@ -0,0 +1,946 @@
...
+* *scopable* means its value depends of the scope (also called channel)
```
Comment (SamirBoulil) [inline STANDARD_FORMAT.md]: depends on

File: STANDARD_FORMAT.md [hunk 5]
```diff
@@ -0,0 +1,946 @@
...
+          1 => array:3 [
+            "locale" => "fr_FR"
+            "scope" => null
+            "data" => "la donnée en français"
+          ]
+        ]
+
+All types of attributes (except the *identifier* and *asset*) can be localisable and/or scopable. In the example above:
+ 
+* there is a localisable image: *a_localizable_image*
```
Comment (SamirBoulil) [inline STANDARD_FORMAT.md]: same

File: STANDARD_FORMAT.md [hunk 6]
```diff
@@ -0,0 +1,946 @@
...
+| text area                    	| string         | `"this is a very very very very very long text"`                                                   |                                                                                                                    |
+| yes/no                       	| boolean        | `true`                                                                                             |                                                                                                                    |
+| metric                       	| array          | `["amount" => "987654321987.123456789123","unit" => "KILOWATT"]`                                   | *amount* and *unit* keys are expected *unit* should be a know unit depending of the metric family of the attribute |
+| price collection             	| array          | `[0 => ["amount" => "45.00","currency" => "USD"], 1 => ["amoun" => "56.53","currency" => "EUR"] ]` | *amount* and *currency* keys are expected for each price *currency* should be a known currency                     |
```
Comment (SamirBoulil) [inline STANDARD_FORMAT.md]: amoun => amount*

File: STANDARD_FORMAT.md [hunk 7]
```diff
@@ -0,0 +1,946 @@
...
+
+
+The following product values data, that represents decimal values, are represented with strings (when `decimal_allowed` is false on attribute properties), in the standard format:
```
Comment (SamirBoulil) [inline STANDARD_FORMAT.md]: when the `decimal_allowed` attribute property is set to false ?

File: STANDARD_FORMAT.md [hunk 8]
```diff
@@ -0,0 +1,946 @@
...
+
+* metric (class *Pim\Component\Catalog\Model\Metric*)
+* price (class *Pim\Component\Catalog\Model\ProductPriceInterface*)
+* number (class *Pim\Component\Catalog\Model\ProductValueInterface*, property *getDecimal*)
+
+They are represented with integer (when `decimal_allowed` is true on attribute properties), in the standard format.
```
Comment (SamirBoulil) [inline STANDARD_FORMAT.md]: with integers* or with an integer

File: STANDARD_FORMAT.md [hunk 9]
```diff
@@ -0,0 +1,946 @@
...
+| unique                 | boolean        | `false`                            |                                                                                                                                                                                        |
+| useable_as_grid_filter | boolean        | `true`                             |                                                                                                                                                                                        |
+| allowed_extensions     | string         | `"pdf,doc"`                        | Extensions separated by commas                                                                                                                                                         |
+| metric_family          | string         | `"Power"`                          | it represents the constant *FAMILY* in classes of *Akeneo/Bundle/MeasureBundle/Family/*                                                                                                |
```
Comment (SamirBoulil) [inline STANDARD_FORMAT.md]: in the classes ?

File: STANDARD_FORMAT.md [hunk 10]
```diff
@@ -0,0 +1,946 @@
...
+The objective is in the future is to use it:
+
+* as versionning format in order to store the history of all entities in the database
+* MAYBE TODO be a base layer for an external REST API
+
+
+---
+
+
+TODO v1:
```
Comment (SamirBoulil) [inline STANDARD_FORMAT.md]: To remove ?

Comment (SamirBoulil) [inline STANDARD_FORMAT.md]: Or maybe we can add a last section '# For the future' or '# next Ideas'
Also maybe we shouldn't reference v2 on the community repo :/

File: STANDARD_FORMAT.md [hunk 11]
```diff
@@ -0,0 +1,946 @@
...
+add a context with [backend_type, attribute_type]
+add asset
+
+TODO v2 ?:
+
+change "data" in price attribute to "amount"
+change "data" in metric attribute to "quantity"
```
Comment (nidup) [inline STANDARD_FORMAT.md]: amount will be used for the first version for both price and metric

Comment (SamirBoulil) [inline STANDARD_FORMAT.md]: Ok

File: STANDARD_FORMAT.md [hunk 12]
```diff
@@ -0,0 +1,928 @@
+# Standard format
+
+The standard format is a normalized array representation of the objects of the PIM. It it used to manipulate (query/update), describe or even sometimes store these objects *inside* the PIM. Currently it is *not* designed to provide a representation of these objects outside the PIM.
+
+The standard format is consistent in term of:
+
+* structure: for instance, products will always be represented the same way
+* data formatting: for instance, dates will always be formatted the same way
+
+The standard format always returns the complete structure, even if data are null.
```
Comment (jjanvier) [inline STANDARD_FORMAT.md]: data IS null

File: STANDARD_FORMAT.md [hunk 13]
```diff
@@ -0,0 +1,928 @@
...
+| enabled       | boolean        | `true`                                                                    |                                                                                                  |
+| values        | array          |                                                                           | see below                                                                                        |
+| created       | string         | `"2016-06-13T00:00:00+02:00"`                                             | formatted to ISO-8601 (see above)                                                                |
+| updated  	    | array          | `"2016-06-13T00:00:00+02:00"`                                             | formatted to ISO-8601 (see above)                                                                |
+| associations  | array          | `["X_SELL" => ["groups" => [0 => "groupA"], "products" => [0 => "foo"]]]` | see below                                                                                        |
+
+
+### Associations
+
+The structure of array is composed like that:
```
Comment (jjanvier) [inline STANDARD_FORMAT.md]: The structure of the array is composed as below:

File: STANDARD_FORMAT.md [hunk 14]
```diff
@@ -0,0 +1,928 @@
...
+            ]
+            "products" => array:1 [
+              0 => "foo"
+            ]
+          ]
+        ]
+
+"X_SELL" represents the *code* of the *Pim\Component\Catalog\Model\AssociationTypeInterface*.
+
+Each element in array "groups" represents the *code* of the *Pim\Component\Catalog\Model\GroupInterface*
```
Comment (jjanvier) [inline STANDARD_FORMAT.md]: in the array

File: STANDARD_FORMAT.md [hunk 15]
```diff
@@ -0,0 +1,928 @@
...
+
+Each element in array "products" represents the *identifier* of the *Pim\Component\Catalog\Model\ProductInterface*
```
Comment (jjanvier) [inline STANDARD_FORMAT.md]: in the array

File: STANDARD_FORMAT.md [hunk 16]
```diff
@@ -0,0 +1,928 @@
...
+| text                         	| string         | `"this is a text"`                                                                                 |                                                                                                                    |
+| text area                    	| string         | `"this is a very very very very very long text"`                                                   |                                                                                                                    |
+| yes/no                       	| boolean        | `true`                                                                                             |                                                                                                                    |
+| metric                       	| array          | `["amount" => "987654321987.123456789123","unit" => "KILOWATT"]`                                   | *amount* and *unit* keys are expected *unit* should be a know unit depending of the metric family of the attribute |
```
Comment (jjanvier) [inline STANDARD_FORMAT.md]: should be a knowN

File: STANDARD_FORMAT.md [hunk 17]
```diff
@@ -0,0 +1,928 @@
...
+| price collection             	| array          | `[0 => ["amount" => "45.00","currency" => "USD"], 1 => ["amount" => "56.53","currency" => "EUR"] ]`| *amount* and *currency* keys are expected for each price *currency* should be a known currency                     |
+
+
+The following product values data, that represents decimal values are represented with strings (when the `decimal_allowed` attribute property is set to false) in the standard format:
+
+* metric (class *Pim\Component\Catalog\Model\Metric*)
+* price (class *Pim\Component\Catalog\Model\ProductPriceInterface*)
+* number (class *Pim\Component\Catalog\Model\ProductValueInterface*, property *getDecimal*)
+
+They are represented with integers (when the `decimal_allowed` attribute property is set to true) in the standard format.
```
Comment (jjanvier) [inline STANDARD_FORMAT.md]: When the decimal_allowed attribute property is set to true, they are represented with integers in the standard format.

File: STANDARD_FORMAT.md [hunk 18]
```diff
@@ -0,0 +1,928 @@
...
+| ---------------------- | -------------- | ---------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| code                   | string         | `"a_date"`                         | it's the identifier of the attribute                                                                                                                                                   |
+| type                   | string         | `"pim_catalog_date"`               |                                                                                                                                                                                        |
+| labels                 | string[]       | `[0 => "A date", 1 => "Une date"]` | each keys of the array represents the *code* of the *Pim\Component\Catalog\Model\LocaleInterface*                                                                                      |
```
Comment (jjanvier) [inline STANDARD_FORMAT.md]: each key

File: STANDARD_FORMAT.md [hunk 19]
```diff
@@ -0,0 +1,928 @@
...
+| negative_allowed       | boolean        | `false`                            |                                                                                                                                                                                        |
+| date_min               | string         | `"2016-09-01T00:00:00+0200"`       | formatted to ISO-8601 (see above)                                                                                                                                                      |
+| date_max               | string         | `"2016-09-01T00:00:00+0200"`       | formatted to ISO-8601 (see above)                                                                                                                                                      |
+| max_file_size          | string         | `"255.00"`                         | limit in Mo                                                                                                                                                                            |
```
Comment (jjanvier) [inline STANDARD_FORMAT.md]: limit in MB

File: STANDARD_FORMAT.md [hunk 20]
```diff
@@ -0,0 +1,928 @@
...
+            "en_US" => "A option"
+          ]
+        ]
+        
+type       | data structure | data example              | notes                                                                                             |
+---------- | -------------- | ------------------------- | ------------------------------------------------------------------------------------------------- |
+code       | string         | `"option_a"`              | it's the identifier of the attribute option                                                       |
+attribute  | string         | `"a_simple_select"`       | the element represents the *code* of the *Pim\Component\Catalog\Model\AttributeInterface*         |
+sort_order | integer        | `0`                       |                                                                                                   |
+labels     | string[]       | `["en_US" => "A option"]` | each keys of the array represents the *code* of the *Pim\Component\Catalog\Model\LocaleInterface* |
```
Comment (jjanvier) [inline STANDARD_FORMAT.md]: each key

Comment (jjanvier) [inline STANDARD_FORMAT.md]: you have it several times

File: STANDARD_FORMAT.md [hunk 21]
```diff
@@ -0,0 +1,928 @@
...
+| currencies       | string[]       | `[0 => "USD", "1 => "EUR"]`                    | each element of the array represents the *code* of the *Pim\Component\Catalog\Model\CurrencyInterface*                                                                                                                                                   |
+| locales          | string[]       | `[0 => "en_US", 1 => "fr_FR"]`                 | each element of the array represents the *code* of the *Pim\Component\Catalog\Model\LocaleInterface*                                                                                                                                                     |
+| category_tree    | string         | `"master"`                                     | only category root. It represents the *code* of the object *Akeneo\Component\Classification\Model\CategoryInterface*                                                                                                                                     |
```
Comment (jjanvier) [inline STANDARD_FORMAT.md]: Root category

File: STANDARD_FORMAT.md [hunk 22]
```diff
@@ -0,0 +1,928 @@
...
+| conversion_units | string[]       |                                                | keys of each element of the array represents the *code* of the *Pim\Component\Catalog\Model\AttributeInterface*. Values of each element of the array represents one of the constant in classes of *Akeneo/Bundle/MeasureBundle/Family/*, except *FAMILY* |
```
Comment (jjanvier) [inline STANDARD_FORMAT.md]: represent
(without S)
2 times

File: STANDARD_FORMAT.md [hunk 23]
```diff
@@ -0,0 +1,928 @@
...
+| priority   | integer        | `100`                                                                                                                                                |                                                                                                                 |
+| conditions | array          | `[0 => ["field" => "a_name", "operator" => "contains", "value" => "description"], 1 =>["field" => "a_text", "operator" => "!=", "value" => "name"]]` | each element of the array represents a condition for *Pim\Component\Catalog\Query\ProductQueryBuilderInterface* |
+| actions    | array          | `[0 => ["type" => "set", "field" => "a_text_area", "value" => "the new description"]]`                                                               | each element of the array represents the action to apply when condition is fulfilled                            |
+
+
+## Usage
+
+The standard format is used:
```
Comment (jjanvier) [inline STANDARD_FORMAT.md]: is used to

File: STANDARD_FORMAT.md [hunk 24]
```diff
@@ -0,0 +1,928 @@
...
+
+* imports
```
Comment (jjanvier) [inline STANDARD_FORMAT.md]: import data

File: STANDARD_FORMAT.md [hunk 25]
```diff
@@ -0,0 +1,928 @@
...
+* exports
```
Comment (jjanvier) [inline STANDARD_FORMAT.md]: export data

File: STANDARD_FORMAT.md [hunk 26]
```diff
@@ -0,0 +1,928 @@
...
+* update objects in memory (imports, PEF for products, Mass Edit)
+* data for PQB filters
```
Comment (jjanvier) [inline STANDARD_FORMAT.md]: define the data expected in the FQCN ProductQueryBuilderInterface filters

**Pull request merged**